src/PluginsManager/Microsoft.Performance.Toolkit.PluginsManager.Core/Manager/PluginsManager.cs

```diff
@@ -113,57 +113,24 @@ public void LoadAdditionalPluginResources(string directory)
             var results = new List<AvailablePlugin>();
             foreach (IPluginDiscoverer discoverer in this.discoverersManager.GetDiscoverersFromSource(source).AsQueryable())
             {
-                IReadOnlyCollection<AvailablePlugin> plugins = await discoverer.DiscoverPluginsLatestAsync(cancellationToken);
-                results.AddRange(plugins);
-            }
-
-            return results;
-        }
-
-        /// <inheritdoc />
-        public async Task<IReadOnlyCollection<AvailablePlugin>> GetAllVersionsOfPlugin(
-            AvailablePlugin availablePlugin,
-            CancellationToken cancellationToken)
-        {
-            Guard.NotNull(availablePlugin, nameof(availablePlugin));
-
-            if (this.PluginSources.Contains(availablePlugin.PluginSource))
-            {
-                foreach (IPluginDiscoverer discoverer in this.discoverersManager.GetDiscoverersFromSource(availablePlugin.PluginSource).AsQueryable())
+                foreach(AvailablePluginInfo info in await discoverer.DiscoverPluginsLatestAsync(cancellationToken))
                 {
-                    IReadOnlyCollection<AvailablePlugin> plugins = await discoverer.DiscoverAllVersionsOfPlugin(
-                        availablePlugin.Identity,
-                        cancellationToken);
-
-                    if (plugins.Any())
-                    {
-                        return plugins;
-                    }
+                    IPluginFetcher fetcher = await GetPluginFetcher(info);
+                    var AvailablePlugin = new AvailablePlugin(info, discoverer, fetcher);
+                    results.Add(AvailablePlugin);
```

Comment:
This should probably be keeping a `Dictionary<PluginIdentity, AvailablePluginInfo> latestVersion` to return the latest versions of a particular plugin found across multiple discoverers (see my other comment about two sources and 3 versions)

Reply:
I see what you are saying. 
I think there's some confusion between discoverer and source though.  For multiple sources, definitely there can be different latest version of a same plugin and we should return all of them. For one particular source that's supported by multiple discoverers, I wonder what it means when different discoverers return different results. 

Noted that the method on L102 gets available plugins from one particular source. The method on L88 gets available plugins from all sources.

Reply:
Yeah, in my example I'm imagining 1 source == 1 discoverer. I think if we ever are in a situation where 1 source has multiple discoverers, we shouldn't even allow adding that source

Reply:
@mslukebo  After thinking about this for a bit - we should still allow multiple discoverers to attempt to discover from one source/feed because there isn't so much `IPluginDiscovererProvider.IsSupportedAsync` can check from a URI. I have updated the logic for handling duplicated within one feed and across multiple feed and added some explanatory comments inline. Can you take another look?